docs: add KB for volumes stuck in attach detach loop when using longhorn on okd (#663)
longhorn/website: content/kb/troubleshooting-volumes-stuck-in-attach-detach-loop-when-using-longhorn-on-okd.md

---
title: "Troubleshooting: Volumes Stuck in Attach/Detach Loop When Using Longhorn on OKD"
author: Jack Lin
draft: false
date: 2023-02-09
categories:
  - "iscsi"
---

## Applicable versions
All Longhorn versions. 

## Symptoms
All volumes stuck in Attach/Detach loop. By using dmesg on storage nodes you can see errors like the following:
```
[Sat Dec 10 18:52:01 2022] audit: type=1400 audit(1670698321.515:7214): avc:  denied  { dac_override } for  pid=231579 comm="iscsiadm" capability=1  scontext=system_u:system_r:iscsid_t:s0 tcontext=system_u:system_r:iscsid_t:s0 tclass=capability permissive=0
[Sat Dec 10 18:52:01 2022] audit: type=1300 audit(1670698321.515:7214): arch=c000003e syscall=83 success=no exit=-13 a0=55b9035185c0 a1=1f8 a2=ffffffffffffff00 a3=0 items=0 ppid=231163 pid=231579 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="iscsiadm" exe="/usr/sbin/iscsiadm" subj=system_u:system_r:iscsid_t:s0 key=(null)
[Sat Dec 10 18:52:01 2022] audit: type=1327 audit(1670698321.515:7214): proctitle=697363736961646D002D6D00646973636F76657279002D740073656E6474617267657473002D700031302E3133312E312E31363
```

## Reason

> Several OKD versions are verified to have this issue.

Reviewer: Looks redundant, suggest removing this line.

Reviewer 2: @ChanYiLin :point_up:

Author: removed

> This is caused by the permissions issue related to SELixux of the host. So the iscsiadm on the host cannot perform correctly.

Reviewer 2: ```suggestion
This is caused by the permissions issue related to the SELixux policies of the host. So the iscsiadm on the host cannot perform correctly if the version is before or equal to `2.1.4` [ref](https://github.com/open-iscsi/open-iscsi/pull/244/commits/6df400925cfa9e723375c6f61524473703054220). 
```

Author: fixed

## Solution

> 1. Upgrade your okd to a newer version which is after `4.12.0-0.okd-2022-12-10`

Reviewer 2: ```suggestion
1. Upgrade your OKD to a newer version which is after `4.12.0-0.okd-2022-12-10` to set the missing policy or the version with a newer version of open-iscsi including the fix. 
```

Author: fixed

> This is caused by the permission issue related to the SELixux policies of the host, so the iscsiadm on the host cannot perform correctly if the version is before or equal to `2.1.4` [ref](https://github.com/open-iscsi/open-iscsi/pull/244/commits/6df400925cfa9e723375c6f61524473703054220). 

Reviewer 3: ```suggestion
This is caused by the permission issue related to the SELinux policies of the host, so the `iscsiadm` on the host cannot perform correctly if the version is before or equal to `2.1.4` [ref](https://github.com/open-iscsi/open-iscsi/pull/244/commits/6df400925cfa9e723375c6f61524473703054220). 
```

Author: thanks!

> This is caused by the permission issue related to the SELixux policies of the host, so the iscsiadm on the host cannot perform correctly if the version is before or equal to `2.1.4` [ref](https://github.com/open-iscsi/open-iscsi/pull/244/commits/6df400925cfa9e723375c6f61524473703054220). 

Reviewer: ```suggestion
Caused by the [permission issue](https://github.com/open-iscsi/open-iscsi/pull/244/commits/6df400925cfa9e723375c6f61524473703054220) related to the host SELinux policies resulting in the `iscsiadm` failing to operate correctly. This issue is likely to happen if the OKD version is before or equal to `2.1.4`
```

Author: Thanks for the wording.
2.1.4 is the iscsi version not OKD version
I will try to refactor the sentence

Author: I have refactored the sentence, please take a look, thanks!

> This is caused by the permissions issue related to the SELixux policies of the host. So the iscsiadm on the host cannot perform correctly if the version is before or equal to `2.1.4` [ref](https://github.com/open-iscsi/open-iscsi/pull/244/commits/6df400925cfa9e723375c6f61524473703054220). 

Reviewer 2: ```suggestion
This is caused by the permission issue related to the SELixux policies of the host. So the iscsiadm on the host cannot perform correctly if the version is before or equal to `2.1.4` [ref](https://github.com/open-iscsi/open-iscsi/pull/244/commits/6df400925cfa9e723375c6f61524473703054220). 
```

Author: fixed

> This is caused by the permissions issue related to the SELixux policies of the host. So the iscsiadm on the host cannot perform correctly if the version is before or equal to `2.1.4` [ref](https://github.com/open-iscsi/open-iscsi/pull/244/commits/6df400925cfa9e723375c6f61524473703054220). 

Reviewer: ```suggestion
This is caused by the permissions issue related to the SELinux policies of the host. So the iscsiadm on the host cannot perform correctly if the version is before or equal to `2.1.4` [ref](https://github.com/open-iscsi/open-iscsi/pull/244/commits/6df400925cfa9e723375c6f61524473703054220). 
```

> 2. Applying `dac_override` using a local CIL via MachineConfig is also a workaround. Please refer to the information in the next section.

Reviewer 2: ```suggestion
2. If in the existing non-working environment, applying `dac_override` using a local CIL via MachineConfig is also a workaround. Please take a look at the below reference links.
```

Author: fixed

> 1. Upgrade your OKD to a newer version which is after `4.12.0-0.okd-2022-12-10` to set the missing policy or the version with a newer version of open-iscsi including the fix.

Reviewer: I am a little confused here. Do you mean there are 2 solutions?
1. Upgrade your OKD to a newer version which is after `4.12.0-0.okd-2022-12-10` + set the missing policy
2. Upgrade to a version with a newer version of open-iscsi including the fix.

Author: No, actually, I think from users' perspective, they don't need to care about the iscsi version.
They just need to upgrade OKD to a newer version to solve the issue.

Reviewer 2: Both are right, but I think we should describe the way @c3y1huang is mentioned. It will make the resolution clear.

Author: upgrade to 4.12.0-0.okd-2022-12-10 will automatically fix the missing policy problem, so users don't have to set the policy by themselves

Author: I have refactored the solution section

> 1. Upgrade your okd to a newer version which is after `4.12.0-0.okd-2022-12-10` to set the missing policy or the version with a newer version of open-iscsi including the fix.

Reviewer 2: ```suggestion
1. Upgrade your OKD to a newer version which is after `4.12.0-0.okd-2022-12-10` to set the missing policy or the version with a newer version of open-iscsi including the fix.
```

Author: fixed

> 2. Upgrade open-iscsi to a newer version including the fix of the permission issue.

Reviewer 2: ```suggestion
2. Upgrade open-iscsi to a newer version including the fix of the permission issue if possible.
```

Author: fixed, thanks!

> 2. If in the existing non-working environment, applying `dac_override` using a local CIL via MachineConfig is also a workaround. Please take a look at the below reference links.

Reviewer: Can we provide specific steps on how to achieve this?

Reviewer 2: This should be fine because it's specific for OKD, we can refer this to the user thread instead because it has been verified there.

Author: I don't have OKD environment now, so I can't verify the steps.
all I can do is to provide the solution from this OKD user's [comment](https://github.com/longhorn/longhorn/issues/4988#issuecomment-1345676772)
I am not familiar with OKD and [MachineConfig](https://docs.okd.io/4.9/post_installation_configuration/machine-configuration-tasks.html#using-machineconfigs-to-change-machines)
From my understanding, the solution is to apply MachineConfig to the OKD and it will start a service with the fix config to apply dac_override to iscsi
But again, I can't verify the steps here.

Reviewer 2: @ChanYiLin It's fine for this KB because it's beyond the scope, but your understanding should be correct.

Author: fixed

> ## Solution
> 
> 1. Upgrade your OKD to a newer version which is after 4.12.0-0.okd-2022-12-10.
> 2. Upgrade open-iscsi to a newer version including the fix for the permission issue.
> 3. If in the existing non-working environment, applying `dac_override` using a local CIL via MachineConfig is also a workaround. Please take a look at the below reference links.

Reviewer 2: ```suggestion
## Solution

There are three ways to resolve the issue. 

1. Upgrade your OKD to a newer version which is after 4.12.0-0.okd-2022-12-10.
2. Upgrade open-iscsi to a newer version including the fix for the permission issue if possible.
3. If in the existing non-working environment, applying `dac_override` using a local CIL via MachineConfig is also a workaround. Please take a look at the below reference links.
```

Author: fixed
